src/v/kafka/requests/fetch_request.cc

```diff
@@ -519,9 +559,128 @@ fetch_api::process(request_context&& rctx, ss::smp_service_group ssg) {
                 [&octx] { return octx.should_stop_fetch(); },
                 [&octx] { return fetch_topic_partitions(octx); });
           })
-          .then(
-            [&octx] { return octx.rctx.respond(std::move(octx.response)); });
+          .then([&octx] { return std::move(octx).send_response(); });
     });
 }
 
+void op_context::reset_context() { initial_fetch = false; }
+
+// decode request and initialize budgets
+op_context::op_context(request_context&& ctx, ss::smp_service_group ssg)
+  : rctx(std::move(ctx))
+  , ssg(ssg)
+  , response_size(0)
+  , response_error(false) {
+    /*
+     * decode request and prepare the inital response
+     */
+    request.decode(rctx);
+    if (likely(!request.topics.empty())) {
+        response.partitions.reserve(request.topics.size());
+    }
+
+    if (auto delay = request.debounce_delay(); delay) {
+        deadline = model::timeout_clock::now() + delay.value();
+    }
+
+    /*
+     * TODO: max size is multifaceted. it needs to be absolute, but also
+     * integrate with other resource contraints that are dynamic within the
+     * kafka server itself.
+     */
+    static constexpr size_t MAX_SIZE = 128 << 20;
```

Comment:
we do not use ALL_CAPS anywhere. I assume this is from old code?

Reply:
also use the `units.h` and use `128_KiB`

Reply:
yes, it was moved from `.h`, will change it.